File: functions/src/features/sorting-ceremony/sorting-ceremony.ts
import {getAuth} from "firebase-admin/auth";
import {logger} from "firebase-functions/v2";
import {getFirestore} from "firebase-admin/firestore";
import {sortingHatAlgorithm} from "./sorting-hat-algorithm";
import {HttpsError, onCall} from "firebase-functions/v2/https";
import {getSignedInUser} from "../../shared/get_signed_in_user";

export type Member = {
  displayName?: string;
  email?: string;
  photoURL?: string;
};

export const sortingCeremony = onCall(
  {region: "europe-west3"},
  async (request) => {
    const loggedUser = await getSignedInUser(request);

    if (loggedUser.customClaims?.house) {
      logger.error("User already has a house");
      throw new HttpsError("already-exists", "User already has a house");
    }

    const housesSnapshot = await getFirestore()
      .collection("houses")
      .get();

    const house = sortingHatAlgorithm(housesSnapshot.docs);

    let member: Member = {
      displayName: loggedUser.displayName,
      email: loggedUser.email,
    };

    if (loggedUser.photoURL) {
      member = {
        ...member,
        photoURL: loggedUser.photoURL,
      };
    }
    await getFirestore()
      .collection("houses")
      .doc(house.id)
      .collection("members")
      .doc(loggedUser.uid)
      .set(member);

    await getAuth().setCustomUserClaims(loggedUser.uid, {
      house: house.id,
      ...loggedUser.customClaims,
    });

    logger.info("Assigned house", house.id, "to user", loggedUser.email);

    return house.id;
  });

File: functions/features/sorting_ceremony/sorting_ceremony.py
import random
from firebase_functions.https_fn import on_call, CallableRequest
from google.cloud.firestore import SERVER_TIMESTAMP
from firebase_admin import auth

from shared.get_signed_in_user import get_signed_in_user
from firestore_client import client as firestore_client
from logger_config import logger
from shared.env import FIREBASE_REGION


@on_call(region=FIREBASE_REGION)
def sorting_ceremony(request: CallableRequest) -> str:
    logged_user = get_signed_in_user(request)

    logged_user_claims = logged_user.custom_claims if logged_user.custom_claims else {}
    user_team = logged_user_claims.get("team", None)

    if user_team is not None:
        logger.info(f"User {logged_user.email} is already in a team: {user_team}")
        return "user_already_in_team"

    teams_data = firestore_client.collection("teams").where('assignable', '==', True).get()
    teams = {team.id: team.get("membersCount") for team in teams_data if team.exists}
    
    sizes =  [teams[team_id] for team_id in teams]
    min_size = min(sizes)
    candidates = [team_id for team_id, size in teams.items() if size == min_size]
    chosen_team = random.choice(candidates)
    
    logger.info(f"User {logged_user.email} assigned to team {chosen_team}")
    
    firestore_client \
        .collection("users") \
        .document(logged_user.uid) \
        .update({
            "team": chosen_team,
            "teamAssignedAt": SERVER_TIMESTAMP
        })

    auth.set_custom_user_claims(logged_user.uid, {**logged_user_claims, "team": chosen_team})

    return chosen_team